# PrichardGarza_ScienceAdvances

## Overview
This code was used to generate figures for Prichard & Garza, et. al., published in Science Advances in 2023. This code generates data structures, graphs, and statistical analyses based on measures from mice exposed to flickering auditory or visual stimuli, and were examined in visual cotex tissue samples from one hemisphere. Tissue analyses including changes in microglia morphology were measured using the Imaris imaging software to reconstruct cells (Figure 1 & Figure 6) and pNFkB colocalization within cells (Figure 5). The other hemisphere of primary visual cortex was examined for cytokine levels (Figure 3 & Figure 7) as well as RNA single cell data (Figure 2 & Figure 4). Final versions for all figures for the publication were edited in Adobe Illustrator.

## Requirements
Matlab 2020B

.csv files including cytokine data

.csv files including microglia morphology data from Imaris

Violin Plot https://github.com/bastibe/Violinplot-Matlab

## Figure/ Code Overview

### Figures 1 and Figure 6
These figures were generated by using matlab code located in the Figure 1 subfolder to create a data structure of morphology measures including branch length, branch depth, process length, process volume, etc. ImageJ data included cell area and number, and requires the GetImageJData.m code. Sholl data was collected using a separate code GetImaris_shollData as this groups the data by indiviudal microglia, then by process radius. Get_Process_Averages.m is used to get avergerages of process data per microglia cell. 
These measures were then compared across groups using ANOVAs then passed to Violinplot functions to create graphs using the graphing_function.m code. Reconstructed microglia images are stored on NeuroMorph.

![image](https://github.com/singerlabgt/PrichardGarza_ScienceAdvances/assets/57195922/5b51318d-9ee7-4ad0-a371-aa7e47827310)

### Figure 4
These figures using RNA sequencing data were generated using R, and can be found in the subfolder.
The transcriptomic data can be found at NCBI GEO (NCBI Gene Expression Omnibus accession number GSE225842 and GSE226822).

![image](https://github.com/singerlabgt/PrichardGarza_ScienceAdvances/assets/57195922/12f22f90-f030-4471-96c0-c35a464897e9)

### Figure 5
The colocalization of pNfkB within neurons or microglia were measured using Imaris software to construct surfaces. Matlab colocalization_code was used to divide the amount of pNFkB within cells by the volume of each cell type, then a ratio was created. Graphs were then created to compare pNFkB volume between cells, then statistically compared using an Anova. 

![image](https://github.com/singerlabgt/PrichardGarza_ScienceAdvances/assets/57195922/364e118b-2cdb-4022-ac72-1d9ba54eb9f4)

### Figure 3 and Figure 7 
These were generated using cytokine data in the attached .csv file.  PLSDA graphs and heatmaps were generated using the PLSDAFunction. Permutations scripts were used to randomly permute the sample names 1k times to create a null distribution to show that the results were not due to chance etc. 

![image](https://github.com/singerlabgt/PrichardGarza_ScienceAdvances/assets/57195922/133753fd-6aaa-469f-b363-b5ce445e1a55)

The scripts also generate individual cytokine bar plots for each of the specified cytokines. These graphs as well as the heatmap are recolored using the folder othercolor. The final colors for the graphs and marks over individual bars to denote significance values were edited using Adobe Illustrator.

![image](https://github.com/singerlabgt/PrichardGarza_ScienceAdvances/assets/57195922/3b46e615-3788-46d1-875a-a129ff4068af)

### Figure 2
This figure of immunity gene sets were generated using the code in the Figure 2 Code and Data folder. Data is included in the .csv file within the folder. 

![image](https://github.com/singerlabgt/PrichardGarza_ScienceAdvances/assets/57195922/02b36e9b-9d4f-4d15-b48f-71061e123048)






